src/node/desktop/src/main/gwt-callback.ts

```diff
@@ -228,9 +228,26 @@ export class GwtCallback extends EventEmitter {
         forceDefaultExtension: boolean,
         focusOwner: boolean,
       ) => {
+        let defaultPath = normalizeSeparatorsNative(resolveAliasedPath(dir));
+
+        // On macOS with Electron 36+, when defaultPath is just a directory and we have a defaultExtension,
+        // provide a complete filename to avoid extension truncation issues (e.g., ".cpp" becoming ".cp")
+        // https://github.com/rstudio/rstudio/issues/16444
+        if (defaultExtension && process.platform === 'darwin') {
+          const endsWithSeparator = defaultPath.endsWith('/');
+          const parsedPath = path.parse(defaultPath);
+          const hasNoFilename = !parsedPath.base || parsedPath.base === '.';
+          const isExistingDir = existsSync(defaultPath) && lstatSync(defaultPath).isDirectory();
```

Comment:
Nit: I think you want `statSync` rather than `lstatSync`, since if this were a symbolic link to a directory, we would want to resolve the symbolic link and check whether that path is a directory or not?